// Copyright 2021 MozoLM Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package mozolm.grpc;

import "mozolm/models/model_config.proto";

option java_package = "com.google.mozolm.grpc";
option java_outer_classname = "GrpcUtilProto";
option java_multiple_files = true;

// Next available ID: 5
message ClientServerConfig {
  // String representing the port where the server will listen.
  string server_port = 1;

  // Type of authorization credentials used in client/server handshake.
  CredentialType credential_type = 2;
  enum CredentialType {
    SSL = 0;
    INSECURE = 1;
  }

  ServerConfig server_config = 3;

  ClientConfig client_config = 4;
}

// Next available ID: 3
message ServerConfig {
  // Model hub configuration.
  ModelHubConfig model_hub_config = 1;

  // Whether the server will wait for client requests.
  bool wait_for_clients = 2;
}

// Next available ID: 6
message ClientConfig {
  // Type of request client will be making to the server.
  ClientRequestType request_type = 2;
  enum ClientRequestType {
    // Client request type not set.
    NOT_SET = 0;

    // Returns k-best items following context_string.
    K_BEST_ITEMS = 1;

    // Randomly generates string.
    RANDGEN = 2;

    // Calculates bits-per-char on given test_corpus.
    BITS_PER_CHAR_CALCULATION = 3;
  }

  // Number of items to return in k_best extraction.
  int32 k_best = 3;

  // Context string prior to k_best extraction or random generation.
  string context_string = 4;

  // Test corpus for calculating bits-per-character.
  string test_corpus = 5;

  // Timeout when waiting for response from server.
  double timeout = 1;
}
